woocommerce/woocommerce-android
# Issue/4661 disconnect button does not work

Parent issue: #4661 (Disconnect button doesn't work from the first time #4718)


### Description
Clicking on the Disconnect button was broken in the `Card reader detail screen`. This PR fixes that.


### Why was it broken?
```   
cardReaderManager.readerStatus.collect { status ->
                when (status) {
                    is Connected -> cardReaderManager.softwareUpdateAvailability.collect(
                        ::handleSoftwareUpdateAvailability
                    )
               }
}
```

In the above code snippet, There are 2 `collect` functions. We know that `collect` is a suspended function. Here is what is happening in the above code.

We start the first suspend `collect` function as soon as the `CardReaderDetailViewModel` starts. Since it is a suspend function, it starts and suspends until the flow terminates. Here's the catch, `readerStatus` is a `StateFlow` and since `StateFlow` is a hot flow. It never terminates. As a result, the second `Collect` suspend function never executes.

Here are the steps to understand it better

1. Go into the `CardDetailFragment` screen --- (First `collect` starts)`readerStatus` will be `Not Connected`
2. You will see `Not connected` state initially
3. Connect to the reader successfully --- `readerStatus` will be `Connected` (We start our second `Collect` suspend function at this point)
4. Now, you will see the option to Disconnect the reader
5. Click on the `Disconnect reader` button. 
6. Nothing happens 💥 

This is because as I said earlier, our first `collect` suspend function has not terminated yet. `readerStatus` is `StateFlow` and since `StateFlow` is a hot flow, it never terminates. As a result, the second `collect` never executes, and hence the `Not Connected` state which gets fired by clicking the Disconnect button has no effect and nothing happens.

### What is the Fix?
Launching the second `collect` suspend function in a separate `Coroutine` fixes this.


### Testing instructions
1. Go into the `CardDetailFragment` screen
2. Click on `Connect card reader`
3. Once connection happens successfully, you should see the option to Disconnect the reader
4. Tap on the `Disconnect reader` button and see that the disconnection actually happens and you get redirected to the `Not Connected` state.


### Images/gif


https://user-images.githubusercontent.com/1331230/131961920-52ec5a79-4064-4802-be70-8a936afe1d54.mp4
